File: libclassicmap_clr/ButtonMappingHeader.cpp
#include "ButtonMappingHeader.h"

#include <vector>
using std::vector;

namespace libclassicmap {
	vector<uint16_t> ButtonMappingHeader::getAdditionalData() {
		vector<uint16_t> vec;

		uint16_t* ptr_from_header_end = (uint16_t*)(this + 1);
		size_t size = (this->_lengthParam - 4) / sizeof(uint16_t);

		for (size_t i = 0; i < size; i++) {
			vec.push_back(ptr_from_header_end[i]);
		}
		return vec;
	}
}

File: ClassicMapLib/ButtonMappingHeader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;

namespace ClassicMapLib {
	[StructLayout(LayoutKind.Explicit, Size = 12)]
	internal struct ButtonMappingHeaderStructure {
		[FieldOffset(2)]
		public ushort classicControllerButton;
		[FieldOffset(10)]
		private ushort _lengthParam;

		public unsafe ushort[] GetAdditionalData() {
			fixed (ButtonMappingHeaderStructure* this_start = &this) {
				void* this_end = this_start + 1;
				ushort* ptr = (ushort*)this_end;
				ushort[] arr = new ushort[(_lengthParam - 4) / sizeof(ushort)];
				for (int i = 0; i < arr.Length; i++) {
					arr[i] = ptr[i];
				}
				return arr;
			}
		}
	}

	public unsafe class ButtonMappingHeader {
		private readonly ButtonMappingHeaderStructure* _ptr;

		public ButtonMappingHeader(void* ptr) {
			_ptr = (ButtonMappingHeaderStructure*)ptr;
		}

		public ClassicControllerButton ClassicControllerButton {
			get {
				return (ClassicControllerButton)_ptr->classicControllerButton;
			}
			set {
				_ptr->classicControllerButton = (ushort)value;
			}
		}

		public IEnumerable<ushort> GetAdditionalData() {
			return _ptr->GetAdditionalData();
		}

		public string GetAdditionalDataAsHexString() {
			return string.Join("", GetAdditionalData().Select(u => u.ToHexString()));
		}
	}
}
